File: 2016/src/day23.ts
function part1(data: string): number {
  return run(data, { a: 7, b: 0, c: 0, d: 0 })
}

function part2(data: string): number {
  return run(data, { a: 12, b: 0, c: 0, d: 0 })
}

function run(data: string, registers: Record<string, number>): number {
  const program = data.split('\n').map((l) => l.split(' '))
  let ip = 0
  while (ip < program.length) {
    // optimize [inc/dec, inc/dec, jnz <reg> -2] sequences
    // these increments/decrements can be converted to addition/subtraction
    const [i0, i1, i2] = program.slice(ip)
    if (
      i2 &&
      i2[0] == 'jnz' &&
      i2[2] == '-2' &&
      ['inc', 'dec'].includes(i0[0]) &&
      ['inc', 'dec'].includes(i1[0]) &&
      (i0[1] == i2[1] || i1[1] == i2[1])
    ) {
      const src = i2[1]
      const dst = i0[1] == src ? i1[1] : i0[1]
      const op = i0[1] == src ? i1[0] : i0[0]
      const mul = op == 'inc' ? 1 : -1
      registers[dst] += mul * Math.abs(registers[src])
      registers[src] = 0
      ip += 3
      continue
    }

    // otherwise, interpret the program normally
    const [op, arg1, arg2] = program[ip++]
    switch (op) {
      case 'cpy': {
        registers[arg2] = arg1 in registers ? registers[arg1] : parseInt(arg1)
        break
      }
      case 'inc': {
        registers[arg1]++
        break
      }
      case 'dec': {
        registers[arg1]--
        break
      }
      case 'jnz': {
        if (registers[arg1] != 0) {
          const jmp = arg2 in registers ? registers[arg2] : parseInt(arg2)
          ip += jmp - 1
        }
        break
      }
      case 'tgl': {
        const mod = ip + registers[arg1] - 1
        if (mod >= 0 && mod < program.length) {
          const code = program[mod]
          code[0] =
            code.length == 2
              ? code[0] == 'inc'
                ? 'dec'
                : 'inc'
              : code[0] == 'jnz'
              ? 'cpy'
              : 'jnz'
        }
      }
    }
  }
  return registers['a']
}

export { part1, part2 }

File: 2017/day23.py
from collections import defaultdict


def part1(data):
    program = [s.split(" ") for s in data.split("\n")]
    regs = defaultdict(int)
    result = 0
    while 0 <= regs["ip"] < len(program):
        code, x, y = program[regs["ip"]]
        if code == "set":
            regs[x] = decode(regs, y)
        elif code == "sub":
            regs[x] -= decode(regs, y)
        elif code == "mul":
            regs[x] *= decode(regs, y)
            result += 1
        elif code == "jnz":
            if decode(regs, x) != 0:
                regs["ip"] += decode(regs, y) - 1
        regs["ip"] += 1
    return result


def part2(data):
    program = [s.split(" ") for s in data.split("\n")]
    regs = defaultdict(int, a=1)
    while 0 <= regs["ip"] < len(program):
        # hook the program at line 11 and
        # implement a faster primality test
        if regs["ip"] == 11:
            if not is_prime(regs["b"]):
                regs["f"] = 0
            regs["d"] = regs["b"]
            regs["e"] = regs["b"]
            regs["g"] = 0
            regs["ip"] = 24

        # otherwise, execute normally
        code, x, y = program[regs["ip"]]
        if code == "set":
            regs[x] = decode(regs, y)
        elif code == "sub":
            regs[x] -= decode(regs, y)
        elif code == "mul":
            regs[x] *= decode(regs, y)
        elif code == "jnz":
            if decode(regs, x) != 0:
                regs["ip"] += decode(regs, y) - 1
        regs["ip"] += 1
    return regs["h"]


def decode(regs, value):
    if value.isalpha():
        return regs[value]
    return int(value)


def is_prime(n):
    if n <= 3:
        return n > 1
    if n % 2 == 0 or n % 3 == 0:
        return False
    i = 5
    while i ** 2 <= n:
        if n % i == 0 or n % (i + 2) == 0:
            return False
        i += 6
    return True
